benchmark/metrics_parser.py

```diff
@@ -0,0 +1,276 @@
+
+    def _scenarios(self):
+        """
+        Return all unique scenarios present in the results.
+        """
+        scenarios = [r['scenario']['name'] for r in self.results]
+        return list(set(scenarios))
+
+    def _filter_results(self, node_count, container_count, version, scenario):
+        """
+        Extract results which match the specified number of nodes,
+        number of containers, Flocker version and benchmarking scenario.
+        """
+        return [r for r in self.results
+                if r['control_service']['node_count'] == node_count
+                and r['control_service']['container_count'] == container_count
+                and r['control_service']['flocker_version'] == version
+                and r['scenario']['name'] == scenario]
+
+    def _create_summary(self):
+        """
+        Summarise the results.
+
+        For each scenario, include the CPU usage for each process,
+        the percentage of containers that converge within a given time
+        limit and the percentage of scenario requests that complete
+        within a given time limit.
+        """
+        summary = defaultdict(list)
+        node_counts = self._node_counts()
+        container_counts = self._container_counts()
+        versions = self._versions()
+        scenarios = self._scenarios()
+        for node, container, version, scenario in itertools.product(
+            node_counts, container_counts, versions, scenarios
+        ):
+            results = self._filter_results(
+                node, container, version, scenario
+            )
+            if len(results) > 0:
+                result = {
+                    u'Flocker Version': version,
+                    u'Nodes': node,
+                    u'Containers': container,
+                    u'Scenario': scenario,
+                    u'Control Service CPU':
+                        cputime_for_process(results, 'flocker-control'),
+                    u'Dataset Agent CPU':
+                        cputime_for_process(results, 'flocker-dataset'),
+                    u'Container Agent CPU':
+                        cputime_for_process(results, 'flocker-contain'),
+                    u'Containers Converged Within Limit':
+                        container_convergence(results, 60),
+                    u'Scenario Requests Within Limit':
+                        request_latency(results, 30),
+                }
+                summary[scenario].append(result)
+        return summary
+
+    def _flatten(self, results):
+        """
+        Flatten a set of results by creating a separate object for each
+        sample in the results.
+        """
+        flattened = []
+        common = dict(
+            [(k, results.get(k)) for k in results.iterkeys() if k != 'samples']
+        )
+
+        metric_type = results['metric']['type']
+
+        for sample in results['samples']:
+            if sample['success']:
+                if metric_type == 'cputime':
```

Comment:
Put the code for cputime and walltime into separate functions that return the `doc` value, and call through a dictionary:
```python
     ...
     flattened.append(METRIC_HANDLER[metric_type](some common args)

METRIC_HANDLER = {
    'cputime': handle_cputime,
    'walltime': handler_walltime,
}
def handle_walltime(some common args):
    doc = dict(common)
    doc['value'] = sample['value']
    return doc
def handle_cputime(some common args):
    ...
```

This will help both with the excessive indentation here, and when we add more metrics.